ament_yamllint/ament_yamllint/main.py

```diff
@@ -0,0 +1,275 @@
+
+    if any(report.values()):
+        return 1
+
+    print('No problems found, checked %d files' % (len(report),))
+    return 0
+
+
+def get_files(paths, extensions, excludes=[]):
+    files = []
+    for path in paths:
+        if os.path.isdir(path):
+            for dirpath, dirnames, filenames in os.walk(path):
+                if 'AMENT_IGNORE' in dirnames + filenames:
+                    dirnames[:] = []
+                    continue
+                # ignore folder starting with . or _
+                dirnames[:] = [d for d in dirnames if d[0] not in ['.', '_']]
+                # ignore excluded folders
+                dirnames[:] = [d for d in dirnames if d not in excludes]
+                dirnames.sort()
+
+                # select files by extension
+                for filename in sorted(filenames):
+                    if filename in excludes:
+                        continue
+                    _, ext = os.path.splitext(filename)
+                    if ext not in ['.%s' % e for e in extensions]:
+                        continue
+                    files.append(os.path.join(dirpath, filename))
+        if os.path.isfile(path):
+            files.append(path)
+    return [os.path.normpath(f) for f in files]
+
+
+def find_executable(file_name, additional_paths=None):
+    path = None
+    if additional_paths:
+        path = os.getenv('PATH', os.defpath)
+        path += os.path.pathsep + os.path.pathsep.join(additional_paths)
+    return shutil.which(file_name, path=path)
+
+
+def invoke_yamllint(files, yamllint_bin=None, config_file=None):
+    if yamllint_bin is None:
+        yamllint_bin = find_executable('yamllint')
+    if not yamllint_bin:
+        raise FileNotFoundError(
+            errno.ENOENT, 'Could not find executable', 'yamllint')
+
+    cmd = [yamllint_bin, '-s', '-f', 'parsable']
+    if config_file:
+        cmd += ['-c', config_file]
+    cmd += files
+
+    report = {f: [] for f in files}
+    try:
+        subprocess.check_output(cmd)
+    except subprocess.CalledProcessError as e:
+        if e.stderr:
+            print(e.stderr.decode())
+        stdout = e.stdout.decode()
+        print(stdout, end='')
+        if not stdout or e.returncode not in (1, 2):
+            raise
+        parser = re.compile(r'^(.+):(\d+):(\d+): \[(.+)\] (.*) \((.+)\)$')
+        for line in stdout.splitlines():
+            m = parser.match(line)
+            if not m:
+                raise ValueError(
+                    'Failed to parse yamllint output: %s' % (line,))
+            try:
+                report[m.group(1)].append({
+                    'line': int(m.group(2)),
+                    'col': int(m.group(3)),
+                    'severity': m.group(4),
+                    'msg': m.group(5),
+                    'id': m.group(6),
+                })
+            except NameError:
+                raise ValueError(
+                    'Got failures for unknown file: %s' % (m.group(1),))
```

Comment:
I'm not a huge fan of doing tons of work inside of an exception handler.  While Python 3 is far better at dealing with this than Python 2 was, it can still be confusing to read the backtrace where multiple exceptions happened inside of one another.  My preferred solution here is to just check what we need to if we intend to re-raise, then save off what we need in the exception handler and continue on.  All other work can be done in the "main" context by checking whether any data was saved from the exception handler.

That said, I don't feel *that* strongly about it, so if you don't think this is worthwhile, I won't block on it.